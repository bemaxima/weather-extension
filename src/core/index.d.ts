interface Location {
	latitude: number;
	longitude: number;
}

interface ILocationProvider {
	getCurrentLocation: () => Promise<Location>;
}

type WeatherData = {
	name: string;
	units: Unit;
	max_temp: number;
	min_temp: number;
	condition: string;
	icon: string;
	temp: number;
}

interface IWeatherForecaster {
	getWeatherInfo: (request: {
		latitude: number,
		longitude: number,
		units: Unit
	}) => Promise<WeatherData>;
}

type Unit = 'celsius' | 'fahrenheit';