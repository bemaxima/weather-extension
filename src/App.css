.app {
  padding: 10px;
  margin: 0 auto;
  text-align: center;
}

@media screen and (-webkit-min-device-pixel-ratio:0) and (min-resolution:.001dpcm) {
  .app {
      width: 200px;
  }
}

.app.loading {
  width: 200px;
  height: 200px;
  background: #c3c3c3;
  display: flex;
  justify-content: center;
  align-items: center;
}

.action-bar {
  display: flex;
  flex-direction: row-reverse;
}

.action-bar .settings {
  padding: 0 2px;
}

.module-header {
  color: #1f1f1f;
  font-weight: 500;
  font-size: 16px;
  text-transform: uppercase;
}

.module-error {
  color: red;
  font-size: 18px;
  font-weight: 600;
  padding: 20px 0;
}

.temp-detail {
  margin-bottom: 12px;
  color: #878787;
}

.weather-icon {
  margin-bottom: 24px;
  height: 70px;
  width: 70px;
}

.temp {
  font-weight: 300;
  line-height: .5;
  font-size: 36px;
  letter-spacing: 1px;
  margin-bottom: 2px;
  white-space: nowrap;
}

.weather {
  color: #878787;
  line-height: 1.14;
  margin-bottom: 16px;
  margin-top: 8px;
}

.loader,
.loader:after {
  border-radius: 50%;
  width: 10em;
  height: 10em;
}
.loader {
  margin: 60px auto;
  font-size: 10px;
  position: relative;
  text-indent: -9999em;
  border-top: 1.1em solid rgba(255, 255, 255, 0.2);
  border-right: 1.1em solid rgba(255, 255, 255, 0.2);
  border-bottom: 1.1em solid rgba(255, 255, 255, 0.2);
  border-left: 1.1em solid #ffffff;
  -webkit-transform: translateZ(0);
  -ms-transform: translateZ(0);
  transform: translateZ(0);
  -webkit-animation: load8 1.1s infinite linear;
  animation: load8 1.1s infinite linear;
}
@-webkit-keyframes load8 {
  0% {
    -webkit-transform: rotate(0deg);
    transform: rotate(0deg);
  }
  100% {
    -webkit-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}
@keyframes load8 {
  0% {
    -webkit-transform: rotate(0deg);
    transform: rotate(0deg);
  }
  100% {
    -webkit-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}