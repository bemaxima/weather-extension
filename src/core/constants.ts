export const LOCALSTORAGE_KEY = 'extension-weather-units';
export const UNITS = {
  Celsius: 'celsius' as Unit,
  Fahrenheit: 'fahrenheit' as Unit
};