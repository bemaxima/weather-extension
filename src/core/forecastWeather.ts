export class ForecastWeather {
	locationProvider: ILocationProvider;
	forecaster: IWeatherForecaster;

	constructor(locationProvider: ILocationProvider, forecaster: IWeatherForecaster) {
		this.locationProvider = locationProvider;
		this.forecaster = forecaster;
	}

	getWeatherInfo(units: Unit) {
		return new Promise<WeatherData>((executor, reject) => {
			this.locationProvider.getCurrentLocation()
				.then(({ latitude, longitude }) => {
					this.forecaster.getWeatherInfo({
						latitude,
						longitude,
						units
					})
						.then(executor);
				})
				.catch(reject);
		});
	}
}