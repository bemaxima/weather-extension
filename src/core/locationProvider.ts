export function getCurrentLocation() {
	return new Promise((executor, reject) => {
		navigator.geolocation.getCurrentPosition(
			({ coords }) => {
				executor({
					latitude: coords.latitude,
					longitude: coords.longitude
				});
			},
			() => {
				reject()
			}
		);
	});
}
