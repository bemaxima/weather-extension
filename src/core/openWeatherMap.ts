import { UNITS } from './constants';

const API_KEY = 'dc4daee96e9f1b3485584c79657f779e';
export const API_URL = `http://api.openweathermap.org/data/2.5/weather?APPID=${API_KEY}`;
export function getWeatherInfo(
	{
		latitude,
		longitude,
		units
	}:
		{
			latitude: number,
			longitude: number,
			units?: Unit
		}) {
	let unitsTemp: string | null = null;
	switch (units) {
		case UNITS.Celsius:
			unitsTemp = 'metric';
			break;
		default:
		case UNITS.Fahrenheit:
			unitsTemp = 'imperial'
			break;
	}
	return new Promise(executor => {
		window
			.fetch(`${API_URL}&lon=${longitude}&lat=${latitude}&units=${unitsTemp}`)
			.then(response => response.json())
			.then(response => {
				const [weather] = response.weather;
				executor({
					temp: response.main.temp,
					max_temp: response.main.temp_max,
					min_temp: response.main.temp_min,
					name: response.name,
					condition: `${weather.main} - ${weather.description}`,
					icon: `https://openweathermap.org/img/wn/${weather.icon}@2x.png`,
					units
				});
			});
	});
}