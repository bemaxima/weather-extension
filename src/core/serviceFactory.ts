import { ForecastWeather } from './forecastWeather';
import * as openWeatherService from './openWeatherMap';
import * as locationService from './locationProvider';

/* 
	In this module we can use any other weather services. 
	Just import the module and pass through `ForecastWeather` class constructor.
*/
export default () => new ForecastWeather(locationService as ILocationProvider, openWeatherService as IWeatherForecaster);