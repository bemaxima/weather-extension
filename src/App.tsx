import React, { useEffect, useReducer, useRef } from 'react';
import './App.css';
import serviceFactory from './core/serviceFactory';
import { UNITS, LOCALSTORAGE_KEY } from './core/constants';
import { reducer } from './stateManagement/reducer';
import * as actionCreators from './stateManagement/actionCreators';
import Error from './components/error';
import Spinner from './components/spinner';
import Viewer from './components/viewer';

function App() {
  const [{ weather, loading, error, units }, dispatch] = useReducer(reducer, {
    weather: {
      temp: null,
      max_temp: null,
      min_temp: null,
      name: 'Unavailable',
      condition: 'Unavailable',
      units: null,
      icon: null
    },
    loading: true,
    error: false,
    units: (localStorage.getItem(LOCALSTORAGE_KEY) || UNITS.Celsius) as Unit
  });

  const service = useRef(serviceFactory());

  function updateWeatherData(units) {
    service.current
      .getWeatherInfo(units)
      .then(response => {
        dispatch(actionCreators.updateWeatherData(response));
      })
      .catch(() => dispatch(actionCreators.errorOccured()));
  }

  function changeUnits(units: Unit) {
    localStorage.setItem(LOCALSTORAGE_KEY, units);
    dispatch(actionCreators.changeUnits(units));
  }

  useEffect(
    () => {
      updateWeatherData(units);
    },
    [units]
  );

  return (
    <div className={`app ${loading ? 'loading' : ''}`}>
      {
        (() => {
          if (error) {
            return <Error />
          }
          else if (loading) {
            return <Spinner />
          }
          else {
            return <Viewer {...weather} onUnitsChange={changeUnits} />
          }
        })()
      }
    </div>
  );
}

export default App;
