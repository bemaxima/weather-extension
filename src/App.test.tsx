import React from 'react';
import { render, act, fireEvent, waitForElementToBeRemoved } from '@testing-library/react';
import App from './App';
import { UNITS, LOCALSTORAGE_KEY } from './core/constants';

const mockGeolocation_v1 = {
  getCurrentPosition: function (succeedCallback) {
    succeedCallback({
      coords: {
        latitude: 2,
        longitude: 1
      }
    });
  },
};

const mockGeolocation_v2 = {
  getCurrentPosition: function (succeedCallback, failureCallback) {
    failureCallback();
  },
};

afterAll(() => {
  window.localStorage.removeItem(LOCALSTORAGE_KEY);
});

describe('App component', () => {

  const SAMPLE_WEATHER_DATA_CELSIUS: any = {
    json: () => ({
      weather: [{
        main: 'Clouds',
        description: 'Partly cloudy'
      }],
      main: {
        temp: 10,
        temp_max: 20,
        temp_min: 5,
      },
      name: 'Unavailable',
      icon: '2dx',
      units: 'celsius'
    })
  };

  const SAMPLE_WEATHER_DATA_FAHRENHEIT: any = {
    json: () => ({
      weather: [{
        main: 'Clouds',
        description: 'Partly cloudy'
      }],
      main: {
        temp: 15,
        temp_max: 30,
        temp_min: 11,
      },
      name: 'Unavailable',
      icon: '2dx',
      units: 'fahrenheit'
    })
  };

  it('shows a complete test scenario which user click on fahrenheit link', async () => {
    (global as any).navigator.geolocation = mockGeolocation_v1;
    let component;
    let fetchGetSpy = jest.spyOn(window, 'fetch').mockResolvedValueOnce(SAMPLE_WEATHER_DATA_CELSIUS);

    await act(async () => {
      component = render(<App />);
      component.getByTestId('loading');
      expect(component.asFragment()).toMatchSnapshot();
    });

    component.getByText('10');
    component.getByText(/20/i);
    component.getByText(/5/i);
    component.getByText('Clouds - Partly cloudy');
    component.getByText('Unavailable');
    expect(component.getByTestId('units').innerHTML).toBe('째C');

    fetchGetSpy = jest.spyOn(window, 'fetch').mockResolvedValueOnce(SAMPLE_WEATHER_DATA_FAHRENHEIT);
    const fahrenheitButton = component.getByTestId('fahrenheit-btn');
    fireEvent.click(fahrenheitButton);

    await act(async () => {
      component = render(<App />);
      expect(component.asFragment()).toMatchSnapshot();
    });

    component.getByTestId('celsius-btn');
    expect(component.getByTestId('units').innerHTML).toBe('째F');
    component.getByText('15');
    component.getByText(/11/i);
    component.getByText(/30/i);
    expect(fetchGetSpy).toHaveBeenCalled();

  });

  it('shows a complete test scenario which user click on celsius link', async () => {
    (global as any).navigator.geolocation = mockGeolocation_v1;
    let component;
    window.localStorage.setItem(LOCALSTORAGE_KEY, UNITS.Fahrenheit);
    let fetchGetSpy = jest.spyOn(window, 'fetch').mockResolvedValueOnce(SAMPLE_WEATHER_DATA_FAHRENHEIT);

    await act(async () => {
      component = render(<App />);
      component.getByTestId('loading');
      expect(component.asFragment()).toMatchSnapshot();
    });

    component.getByText('15');
    component.getByText(/30/i);
    component.getByText(/11/i);
    component.getByText('Clouds - Partly cloudy');
    component.getByText('Unavailable');
    expect(component.getByTestId('units').innerHTML).toBe('째F');

    fetchGetSpy = jest.spyOn(window, 'fetch').mockResolvedValueOnce(SAMPLE_WEATHER_DATA_CELSIUS);

    const celsiusButton = component.getByTestId('celsius-btn');
    fireEvent.click(celsiusButton);

    await act(async () => {
      component = render(<App />);
      expect(component.asFragment()).toMatchSnapshot();
    });

    component.getByTestId('fahrenheit-btn');
    expect(component.getByTestId('units').innerHTML).toBe('째C');
    component.getByText('10');
    component.getByText(/20/i);
    component.getByText(/5/i);
    expect(fetchGetSpy).toHaveBeenCalled();
  });

  it('render in a state that user does not allow browser to acquire location info', async () => {
    (global as any).navigator.geolocation = mockGeolocation_v2;
    let component;
    window.localStorage.setItem(LOCALSTORAGE_KEY, UNITS.Celsius);

    await act(async () => {
      component = render(<App />);
      component.getByTestId('loading');
      expect(component.asFragment()).toMatchSnapshot();
    });
    component.getByText('Location is not available.');
  });
});

