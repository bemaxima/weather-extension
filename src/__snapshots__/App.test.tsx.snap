// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component render in a state that user does not allow browser to acquire location info 1`] = `
<DocumentFragment>
  <div
    class="app loading"
  >
    <div
      class="loader"
      data-testid="loading"
    />
  </div>
</DocumentFragment>
`;

exports[`App component shows a complete test scenario which user click on celsius link 1`] = `
<DocumentFragment>
  <div
    class="app loading"
  >
    <div
      class="loader"
      data-testid="loading"
    />
  </div>
</DocumentFragment>
`;

exports[`App component shows a complete test scenario which user click on celsius link 2`] = `
<DocumentFragment>
  <div
    class="app loading"
  >
    <div
      class="loader"
      data-testid="loading"
    />
  </div>
</DocumentFragment>
`;

exports[`App component shows a complete test scenario which user click on fahrenheit link 1`] = `
<DocumentFragment>
  <div
    class="app loading"
  >
    <div
      class="loader"
      data-testid="loading"
    />
  </div>
</DocumentFragment>
`;

exports[`App component shows a complete test scenario which user click on fahrenheit link 2`] = `
<DocumentFragment>
  <div
    class="app loading"
  >
    <div
      class="loader"
      data-testid="loading"
    />
  </div>
</DocumentFragment>
`;
